vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt

```diff
@@ -0,0 +1,236 @@
+        }
+
+        when (val asyncUsers = currentState.directoryUsers) {
+            is Uninitialized -> {
+            }
+            is Loading -> renderLoading()
+            is Fail -> renderFailure(asyncUsers.error)
+            is Success -> buildDirectoryUsers(
+                    asyncUsers(),
+                    currentState.searchTerm,
+                    emptyList()
+            )
+        }
+    }
+
+    private fun buildlocalContacts(currentState: TchapContactListViewState) {
+        val userList = currentState.filteredRoomSummaries.mapNotNull { item ->
+            item.directUserId?.let {
+                User(it, item.displayName, item.avatarUrl)
+            }
+        }
+        val tchapContactList: MutableList<Any> = userList.toMutableList()
+
+        tchapContactList.addAll(currentState.filteredLocalContacts.filter { mappedContact ->
+            mappedContact.emails.any { it.matrixId != null && !userList.any { user -> user.userId == it.matrixId } }
+        })
+
+        userListHeaderItem {
+            id("local_header", 1)
+            header(stringProvider.getString(R.string.local_address_book_header))
+        }
+
+        tchapContactList.sortWith(object : Comparator<Any> {
+            override fun compare(contact1: Any, contact2: Any): Int {
+                val lhs = when (contact1) {
+                    is User          -> contact1.getBestName()
+                    is MappedContact -> contact1.displayName
+                    else             -> null
+                }
+
+                val rhs = when (contact2) {
+                    is User          -> contact2.getBestName()
+                    is MappedContact -> contact2.displayName
+                    else             -> null
+                }
+
+                if (lhs == null) {
+                    return -1
+                } else if (rhs == null) {
+                    return 1
+                }
+
+                return String.CASE_INSENSITIVE_ORDER.compare(lhs, rhs)
+            }
+        })
+
+        tchapContactList.forEach { item ->
+            when (item) {
+                is User -> {
+                    userDirectoryUserItem {
+                        id(item.userId)
+                        selected(false)
+                        matrixItem(item.toMatrixItem())
+                        avatarRenderer(avatarRenderer)
+                        clickListener { _ ->
+                            callback?.onItemClick(item)
+                        }
+                    }
+                }
+
+                is MappedContact -> {
+                    contactItem {
+                        id(item.id)
+                        mappedContact(item)
+                        avatarRenderer(avatarRenderer)
+                    }
+                    item.emails
+                            .forEachIndexed { index, it ->
+                                if (it.matrixId == null) return@forEachIndexed
```

Comment:
This should work also:
`it.matrixId ?: return@forEachIndexed`

Reply:
Yes, this code comes from element. The result is the same.